src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+
+                var valueLow = value;
+                var valueLowLength = n;
+                var valueHigh = value + n;
+                var valueHighLength = valueLength - n;
+
+                var bitsLow = bits;
+                var bitsLowLength = n * 2;
+                var bitsHigh = bits + n * 2;
+                var bitsHighLength = bitsLength - n * 2;
+
+                Square(valueLow, valueLowLength,
+                       bitsLow, bitsLowLength);
+                Square(valueHigh, valueHighLength,
+                       bitsHigh, bitsHighLength);
+
+                var fold = AddFold(valueHigh, valueHighLength,
+                                   valueLow, valueLowLength);
+                var foldLength = fold[fold.Length - 1] == 0
+                               ? fold.Length - 1 : fold.Length;
+
+                fixed (uint* core = Square(fold, foldLength))
+                {
+                    var coreLength = foldLength * 2;
+                    if (core[coreLength - 1] == 0)
+                        coreLength = coreLength - 1;
+
+                    SubtractCore(bitsHigh, bitsHighLength,
+                                 bitsLow, bitsLowLength,
+                                 core, coreLength);
+
+                    // merge the result
+                    AddSelf(bits + n, bitsLength - n, core, coreLength);
+                }
+            }
+        }
+
+        public static uint[] Multiply(uint left, uint right)
+        {
+            var bits = new uint[2];
+
+            // multiplies the bits
+            var digits = (ulong)left * right;
+            bits[0] = (uint)digits;
+            bits[1] = (uint)(digits >> 32);
+
+            return bits;
+        }
+
+        public static uint[] Multiply(uint[] left, uint right)
+        {
+            Debug.Assert(left != null);
+
+            var i = 0;
+            var carry = 0UL;
+            var bits = new uint[left.Length + 1];
+
+            // multiplies the bits
+            for (; i < left.Length; i++)
+            {
+                var digits = (ulong)left[i] * right + carry;
+                bits[i] = (uint)digits;
+                carry = digits >> 32;
+            }
+            bits[i] = (uint)carry;
+
+            return bits;
+        }
+
+        public static uint[] Multiply(uint[] left, uint[] right)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(right != null);
+
+            if (left.Length < right.Length)
+                return Multiply(right, right.Length, left, left.Length);
+            return Multiply(left, left.Length, right, right.Length);
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static uint[] Multiply(uint[] left, int leftLength,
+                                              uint[] right, int rightLength)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(right != null);
+            Debug.Assert(left.Length >= leftLength);
```

Comment:
Consider checking the precondition `leftLength >= rightLength` here as well as in the overload below.